# General Comments about the data set
This table provides the volume of violent crime (murder and nonnegligent manslaughter, rape, robbery, and aggravated assault) and property crime (burglary, larceny-theft, and motor vehicle theft) as reported by city and town law enforcement agencies (listed alphabetically by state) that contributed data to the UCR Program. (Note:  Arson is not included in the property crime total in this table; however, if complete arson data were provided, they will appear in the arson column.)

Population estimation
For the 2013 population estimates used in this table, the FBI computed individual rates of growth from one year to the next for every city/town and county using 2010 decennial population counts and 2011 through 2012 population estimates from the U.S. Census Bureau. Each agency’s rates of growth were averaged; that average was then applied and added to its 2012 Census population estimate to derive the agency’s 2013 population estimate.

 

# Instructions

We'll use the property crime model you've been working on with, based on the FBI:UCR data. Since your model formulation to date has used the entire New York State 2013 dataset, you'll need to validate it using some of the other crime datasets available at the FBI:UCR website. Options include other states crime rates in 2013 or crime rates in New York State in other years or a combination of these.
__I chose to test the model that I trained on NY to predict what will happen in CA, and see if the model is a good fit for the new data set__

Iterate
Based on the results of your validation test, create a revised model, and then test both old and new models on a new holdout or set of folds.

Include your model(s) and a brief writeup of the reasoning behind the validation method you chose and the changes you made to submit and review with your mentor.

In [106]:
# import all the modules I need
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
pd.set_option('mode.chained_assignment', None)
%matplotlib inline
import seaborn as sns
import math


# import in the data table I downloaded from UCR
df = pd.DataFrame()
df = pd.read_excel(
    "table_8_offenses_known_to_law_enforcement_new_york_by_city_2013.xls", sheet_name='13tbl8ny')

Need to clean by:
- fixing headers (line 3) and deleting rows 0-2
- dropping Rape\n(revised\ndefinition)1 column 
- filling in NaN with 0


Need to create features by:
- changing categorical variables to 0 and 1

In [107]:
df.drop([0, 1, 2], inplace=True)

In [108]:
df.drop('Unnamed: 4', axis=1, inplace=True)

In [109]:
df.columns = df.loc[3].values

In [110]:
df.drop([3, 352, 353, 354], inplace=True)

In [111]:
df['Arson3'] = df['Arson3'].fillna(0)
df

City Population Violent\ncrime  \
4                 Adams Village       1861              0   
5      Addison Town and Village       2577              3   
6                 Akron Village       2846              3   
7                        Albany      97956            791   
8                Albion Village       6388             23   
9                Alfred Village       4089              5   
10             Allegany Village       1781              3   
11                 Amherst Town     118296            107   
12           Amityville Village       9519              9   
13                    Amsterdam      18182             30   
14               Arcade Village       2053              0   
15              Ardsley Village       4523              5   
16            Asharoken Village        658              0   
17               Attica Village       2522              2   
18                       Auburn      27270             96   
19                 Avon Village       3333              1   
20        Baldwinsville Village       7473              5   
21         Ballston Spa Village       5418              8   
22                      Batavia      15374             57   
23                 Bath Village       5733             15   
24                 Bedford Town      17627              5   
25               Bethlehem Town      34243             13   
26                   Binghamton      46304            284   
27                  Black River       1410              0   
28          Blooming Grove Town      12001              8   
29              Bolivar Village       1022              1   
30                  Bolton Town       2325              2   
31            Boonville Village       2054              1   
32                   Brant Town       2065              4   
33                     Brewster       2358              0   
..                          ...        ...            ...   
322         Tupper Lake Village       3642              3   
323         Tuxedo Park Village        615              0   
324                 Ulster Town      12195             14   
325                       Utica      61686            361   
326                 Vestal Town      27980             16   
327              Walden Village       6887             18   
328               Wallkill Town      27940             40   
329              Walton Village       3012              6   
330              Warsaw Village       3453              6   
331     Washingtonville Village       5828              2   
332  Waterford Town and Village       8392              4   
333            Waterloo Village       5175             11   
334                   Watertown      28179            109   
335                  Watervliet      10245             26   
336        Watkins Glen Village       1862              6   
337             Waverly Village       4337              5   
338    Webster Town and Village      43777             21   
339           Weedsport Village       1797              0   
340          Wellsville Village       4574             11   
341   Westhampton Beach Village       1732              2   
342            West Seneca Town      44821             35   
343           Whitehall Village       2604              6   
344                White Plains      57559             78   
345          Whitesboro Village       3733              4   
346             Whitestown Town       9141              0   
347               Woodbury Town      10685              3   
348           Woodridge Village        829              7   
349              Woodstock Town       5931              2   
350                     Yonkers     199134           1036   
351               Yorktown Town      36643             15   

    Murder and\nnonnegligent\nmanslaughter Rape\n(legacy\ndefinition)2  \
4                                        0                           0   
5                                        0                           0   
6                     

In [112]:
df['Population'] = df['Population'].astype(int)
df['Population2'] = df['Population']**2

In [113]:
df['Robbery'] = df['Robbery'].astype(int)
df['Robbery'] = df['Robbery'].apply(lambda x: (x > 0)*1)

In [114]:
df['Murder and\nnonnegligent\nmanslaughter'] = df['Murder and\nnonnegligent\nmanslaughter'].astype(
    int)
df['Murder and\nnonnegligent\nmanslaughter'] = df['Murder and\nnonnegligent\nmanslaughter'].apply(
    lambda x: (x > 0)*1)

In [115]:
from sklearn import linear_model
%matplotlib inline
pd.options.display.float_format = '{:.3f}'.format

# Suppress annoying harmless error.
import warnings
warnings.filterwarnings(action="ignore", module="scipy",
                        message="^internal gelsd")

In [116]:
regr = linear_model.LinearRegression()
regr.fit(df[['Population', 'Population2',
             'Murder and\nnonnegligent\nmanslaughter', 'Robbery']], df['Property\ncrime'])

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [117]:
regr.coef_

array([ 1.69008467e-02, -1.73980871e-07,  6.15497194e+02,  1.48320688e+02])

In [118]:
regr.score(df[['Population', 'Population2',
               'Murder and\nnonnegligent\nmanslaughter', 'Robbery']], df['Property\ncrime'])

0.9935839631191373

Predictions. First figuring out how many property crimes there would be in Davis with the specification we had earlier.

With Drew, we realized how important it was for population to be in the regression specification. When we took out population, the regr score went way down. We see how adding random noise increases the regr score by overfitting the data. 

In [119]:
# predicting the nuber of property crimes in Davis:
regr.predict([[68111, 68111**2, 0, 0]])

array([318.62882533])

In [120]:
regr = linear_model.LinearRegression()
regr.fit(df[['Population2', 'Murder and\nnonnegligent\nmanslaughter',
             'Robbery']], df['Property\ncrime'])
regr.score(df[['Population2', 'Murder and\nnonnegligent\nmanslaughter',
               'Robbery']], df['Property\ncrime'])

0.05208377602595937

In [121]:
regr = linear_model.LinearRegression()
X = df[['Population2', 'Murder and\nnonnegligent\nmanslaughter', 'Robbery']].values
N = np.random.randn(X.shape[0], 10)
X = np.concatenate((X, N), axis=1)
regr.fit(X, df['Property\ncrime'])

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [122]:
regr.score(X, df['Property\ncrime'])

0.07300583209075984

Now, let's use the new linear regression code that I learned in this lesson to refit the model and check params, pvalues, and the rsquared value for NY

In [123]:
import statsmodels.formula.api as smf
from statsmodels.sandbox.regression.predstd import wls_prediction_std

In [125]:
#rename and change the data type to be numerical instead of an object
df['Property\ncrime'] = df['Property\ncrime'].astype(int)
df['Population2'] = df['Population2'].astype(int)
df['Property_crime'] = df['Property\ncrime']
df['Murder'] = df['Murder and\nnonnegligent\nmanslaughter']

In [93]:
# Re-fit the model here.
linear_formula = 'Property_crime ~ Population + Murder'

# Fit the model to our data using the formula.
lm = smf.ols(formula=linear_formula, data=df).fit()
print(lm.params)
print(lm.pvalues)
print(lm.rsquared)
print(lm.conf_int())

Intercept     24.873
Population     0.017
Murder       654.243
dtype: float64
Intercept    0.491
Population   0.000
Murder       0.000
dtype: float64
0.9934062182812436
                 0       1
Intercept  -46.097  95.844
Population   0.017   0.017
Murder     462.613 845.873


In [71]:
# import in the data table I downloaded for CA
df_CA = pd.DataFrame()
df_CA = pd.read_excel(
    'table_8_offenses_known_to_law_enforcement_california_by_city_2013.xls', sheet_name='13tbl8ca')

In [72]:
df_CA.drop([0, 1, 2], inplace=True)

In [73]:
df_CA.drop('Unnamed: 4', axis=1, inplace=True)

In [74]:
df_CA.columns = df_CA.loc[3].values
df_CA

City  Population  \
3                                                 City  Population   
4                                             Adelanto       31165   
5                                         Agoura Hills       20762   
6                                              Alameda       76206   
7                                               Albany       19104   
8                                             Alhambra       84710   
9                                          Aliso Viejo       50005   
10                                             Alturas        2681   
11                                     American Canyon       20068   
12                                             Anaheim      345320   
13                                            Anderson       10098   
14                                             Antioch      106447   
15                                        Apple Valley       71159   
16                                             Arcadia       57855   
17                                              Arcata       17805   
18                                       Arroyo Grande       17634   
19                                             Artesia       16734   
20                                               Arvin       20390   
21                                          Atascadero       28938   
22                                            Atherton        7278   
23                                             Atwater       28906   
24                                              Auburn       13779   
25                                              Avalon        3777   
26                                              Avenal       14712   
27                                               Azusa       47754   
28                                         Bakersfield      361859   
29                                        Baldwin Park       76745   
30                                             Banning       30503   
31                                             Barstow       23144   
32                                         Bear Valley        5328   
..                                                 ...         ...   
438                                              Vista       96712   
439                                             Walnut       30279   
440                                       Walnut Creek       66149   
441                                          Waterford        8590   
442                                        Watsonville       52076   
443                                               Weed        2927   
444                                        West Covina      107867   
445                                     West Hollywood       34902   
446                                   Westlake Village        8440   
447                                        Westminster       91885   
448                                        Westmorland        2279   
449                                    West Sacramento       49751   
450                                          Wheatland        3522   
451                                           Whittier       86450   
452                                           Wildomar       33453   
453                                           Williams        5176   
454                                            Willits        4832   
455                                            Willows        6122   
456                                            Windsor       27249   
457                                            Winters        6992   
458                                           Woodlake        7417   
459                                           Woodland       56523   
460                                        Yorba Linda       67492   
461                                         Yountville        2969   
462                                              Yreka        7639   
463                                          Yuba City      

In [75]:
df_CA.drop([3, 466, 467], inplace=True)

In [77]:
df_CA['Murder'] = df_CA['Murder and\nnonnegligent\nmanslaughter']
df_CA['Property_crime'] = df_CA['Property\ncrime']

In [78]:
df_CA['Murder'] = df_CA['Murder'].astype(int)
df_CA['Murder'] = df_CA['Murder'].apply(lambda x: (x > 0)*1)

In [85]:
df_CA['Population'] = df_CA['Population'].astype(int)
df_CA['Property_crime'] = df_CA['Property_crime'].astype(int)

In [94]:
# Re-fit the model here to CA 
linear_formula = 'Property_crime ~ Population + Murder'

# Fit the model to our data using the formula.
lm_CA = smf.ols(formula=linear_formula, data=df_CA).fit()
print(lm_CA.params)
# print(lm.pvalues)
# print(lm.rsquared)
# print(lm.conf_int())

Intercept    -57.299
Population     0.024
Murder       589.797
dtype: float64


In [102]:
y_pred = lm.predict(df_CA[['Property_crime', 'Population', 'Murder']])
y = df_CA['Property_crime']
SS_Residual = sum((y - y_pred) ** 2)
SS_Total = sum((y - np.mean(y)) ** 2)
r_squared = 1 - (float(SS_Residual)) / SS_Total
print(r_squared)
print(lm.pvalues)

0.8032009799627111
Intercept    0.491
Population   0.000
Murder       0.000
dtype: float64


We used the linear model that was trained on the NY data set from 2013 and tried it on the 2013 CA data set. This gave us a R^2 value of 0.803.

Next, let's validate our model by comparing how the CA trained linear model will work on the NY data set.

For my model to be trustworthy, I expect the R^2 values to be the same or very similar.

In [103]:
y_pred = lm_CA.predict(df[['Property_crime', 'Population', 'Murder']])
y = df['Property_crime']
SS_Residual = sum((y - y_pred) ** 2)
SS_Total = sum((y - np.mean(y)) ** 2)
r_squared = 1 - (float(SS_Residual)) / SS_Total
print(r_squared)
print(lm_CA.pvalues)

0.7998327830081634
Intercept    0.617
Population   0.000
Murder       0.001
dtype: float64


Yay! The R^2 value is 0.799, very similar to the 0.803 R^2 value from before. So, I conclude that my model is trustworthy.

Future steps:
- apply the NY 2013 linear model I made to another year, say 2015, for NY
- I expect that R^2 values would be closer or improved

- do the fold analysis with 5 groups/folds to see if the components listed at the end give similar values and then I can see whether or not I'm overfitting or not.